src/gf.c

```diff
@@ -177,53 +224,187 @@ void jl_mk_builtin_func(jl_datatype_t *dt, const char *name, jl_fptr_t fptr)
+    jl_typemap_visitor(gf->name->mt->cache, fptr, (void*)replaced);
+}
+
+struct set_world {
+    jl_method_instance_t *replaced;
+    size_t world;
+};
+static int set_max_world2(jl_typemap_entry_t *entry, void *closure0)
+{
+    struct set_world *closure = (struct set_world*)closure0;
+    // entry->max_world should be <= closure->replaced->max_world and >= closure->world
+    if (entry->func.linfo == closure->replaced) {
+        entry->max_world = closure->world;
+    }
+    return 1;
+}
+
+
+JL_DLLEXPORT jl_method_instance_t* jl_set_lambda_inferred(
+        jl_method_instance_t *li, jl_value_t *rettype,
+        jl_value_t *const_api, jl_value_t *inferred,
+        size_t min_world, size_t max_world)
+{
+    JL_GC_PUSH1(&li);
+    assert(min_world <= max_world && "attempting to set invalid world constraints");
+    assert(li->inInference && "shouldn't be caching an inference result for a MethodInstance that wasn't being inferred");
+    if (min_world != li->min_world || max_world != li->max_world) {
+        if (li->def == NULL) {
+            // thunks don't have multiple references, so just update in-place
+            li->min_world = min_world;
+            li->max_world = max_world;
+        }
+        else {
+            JL_LOCK(&li->def->writelock);
+            assert(min_world >= li->def->min_world && max_world <= li->def->max_world);
+            int uninferred =  jl_is_uninferred(li);
+            if (uninferred && li->min_world >= min_world && li->max_world <= max_world) {
+                // expand the current (uninferred) entry to cover the full inferred range
+                if (li->min_world != min_world) {
+                    li->min_world = min_world;
+                    jl_typemap_visitor(li->def->specializations, set_min_world, (void*)li);
+                }
+                if (li->max_world != max_world) {
+                    li->max_world = max_world;
+                    jl_typemap_visitor(li->def->specializations, set_max_world, (void*)li);
+                }
+            }
+            else {
+                // clip applicability of old method instance (uninferred or inferred)
+                // to make it easier to find the inferred method
+                // (even though the real applicability was unchanged)
+                // there are 6(!) regions here to consider + boundary conditions for each
+                if (li->max_world >= min_world && li->min_world <= max_world) {
+                    // there is a non-zero overlap between [li->min, li->max] and [min, max]
+                    // there are now 4 regions left to consider
+                    // TODO: also take into account li->def->world range when computing preferred division
+                    if (li->max_world > max_world) {
+                        // prefer making it applicable to future ages,
+                        // as those are more likely to be useful
+                        li->min_world = max_world + 1;
```

Comment:
TODO: can't change this, as it might trick someone into thinking that the new inferred range fully covers the range that it was previously cached for